docker-images/syntax-highlighter/crates/scip-syntax/src/locals.rs

```diff
@@ -1,504 +1,721 @@
-                scope_modifier,
-                reassignment_behavior,
+                current_scope = self.parent(current_scope)
+            }
+            // Before adding the new scope we first attach all
+            // definitions that belong to the current scope
+            self.add_defs_while(current_scope, &mut definitions_iter, |def_capture| {
+                def_capture.node.start_byte() < scope.start_byte()
             });
-        } else if let Some(group) = reference {
-            let identifier = match node.utf8_text(source_bytes) {
-                Ok(identifier) => identifier,
-                Err(_) => continue,
-            };
-
-            references.push(Reference {
-                range: node.into(),
-                group,
-                identifier,
-                node,
+            self.add_refs_while(current_scope, &mut references_iter, |ref_capture| {
+                ref_capture.node.start_byte() < scope.start_byte()
             });
-        } else {
-            let scope = match scope {
-                Some(scope) => scope,
-                None => continue,
-            };
 
-            scopes.push(Scope::new(scope.node));
+            let new_scope =
+                self.arena
+                    .alloc(Scope::new(scope_ty.to_string(), scope, Some(current_scope)));
+            self.get_scope_mut(current_scope).children.push(new_scope);
+            current_scope = new_scope
         }
-    }
 
-    let mut root = Scope::new(root_node);
+        // We need to climb back to the top level scope and add
+        // all remaining definitions
+        loop {
+            let scope_end_byte = self.end_byte(current_scope);
+            self.add_defs_while(current_scope, &mut definitions_iter, |def_capture| {
+                def_capture.node.start_byte() < scope_end_byte
+            });
+            self.add_refs_while(current_scope, &mut references_iter, |ref_capture| {
+                ref_capture.node.start_byte() < scope_end_byte
+            });
 
-    // Sort smallest to largest, so we can pop off the end of the list for the largest, first scope
-    scopes.sort_by_key(|m| {
-        (
-            std::cmp::Reverse(m.range.start_line),
-            m.range.end_line - m.range.start_line,
-            m.range.end_col - m.range.start_col,
-        )
-    });
+            if current_scope == top_scope {
+                break;
+            }
 
-    let capacity = lvalues.len() + references.len();
+            current_scope = self.parent(current_scope)
+        }
 
-    // Add all the scopes to our tree
-    while let Some(m) = scopes.pop() {
-        root.insert_scope(m);
+        assert!(
+            definitions_iter.next().is_none(),
+            "Should've entered all definitions into the tree"
+        );
     }
 
-    while let Some(m) = lvalues.pop() {
-        root.insert_lvalue(m);
+    /// Walks up the scope tree and tries to find the definition for a given name
+    fn find_def(
```

Comment:
nit: Since this is fallible and returns an `Option` on failure, I think it would be more idiomatic to have a `try_` prefix.

```suggestion
    fn try_find_def(
```

Reply:
`vec![].iter().find()` also returns an Option, right? I'd agree if there _also_ was a non-fallible find function around here.

Reply:
Fair point. Funnily enough, I did find a separate `try_find` on Iterator here: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_find